common/web/gesture-recognizer/src/paddedZoneSource.ts

```diff
@@ -0,0 +1,68 @@
+namespace com.keyman.osk {
+
+  export class PaddedZoneSource implements RecognitionZoneSource {
+    private readonly root: RecognitionZoneSource;
+
+    /**
+     * [0]: y (top)
+     * [1]: x (left)
+     * [2]: height (top+bottom)
+     * [3]: width (left+right)
+     */
+    private edgePadding: number[];
+
+    // Positive values 'shrink' the new zone compared to the old zone, while negative ones
+    // 'expand' it instead.
+    constructor(rootZoneSource: RecognitionZoneSource, ...edgePadding: number[]) {
+      this.root = rootZoneSource;
+      // In case it isn't yet defined.
+      edgePadding = edgePadding || [0, 0, 0, 0];
+
+      if(typeof edgePadding == 'number') {
+        this.edgePadding = [edgePadding, edgePadding, 2 * edgePadding, 2 * edgePadding];
+        return;
+      } else {
+        // Modeled after CSS styling definitions... just with preprocessed numbers, not strings.
+        switch(edgePadding.length) {
+          case 1:
+            // all sides equal
+            const val = edgePadding[0];
+            this.edgePadding = [val, val, 2 * val, 2 * val];
+            break;
+          case 2:
+            // top & bottom, left & right
+            const yVal = edgePadding[0];
+            const xVal = edgePadding[1];
+            this.edgePadding = [yVal, xVal, 2 * yVal, 2 * xVal];
+            break;
+          case 3:
+            // top, left & right, bottom
+            this.edgePadding = [edgePadding[0],
+                                edgePadding[1],
+                                edgePadding[0] + edgePadding[2],
+                                2*edgePadding[1]];
+          case 4:
+            // top, right, bottom, left
+            this.edgePadding = [edgePadding[0],
+                                edgePadding[3], // we want the `left` entry internally, not the `right`.
+                                edgePadding[0] + edgePadding[2],
+                                edgePadding[1] + edgePadding[3]];
```

Comment:
So the input of this function is top, right, bottom, left, but the internal member is top, left, height, width? Seems likely to cause us to trip up in the future? (But if we change the internal member to an object as I suggest, this problem goes away.)